src/concurrency/ts_order_txn_manager.cpp

```diff
@@ -134,107 +228,186 @@ bool TsOrderTxnManager::PerformInsert(const ItemPointer &location) {
 void TsOrderTxnManager::PerformUpdate(const ItemPointer &location) {
   oid_t tile_group_id = location.block;
   oid_t tuple_id = location.offset;
 
   auto &manager = catalog::Manager::GetInstance();
   auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
 
-  //  PL_ASSERT(tile_group_header->GetTransactionId(tuple_id) ==
-  //         current_txn->GetTransactionId());
-  //  PL_ASSERT(tile_group_header->GetEndCommitId(tuple_id) == MAX_CID);
+  PL_ASSERT(tile_group_header->GetTransactionId(tuple_id) ==
+         current_txn->GetTransactionId());
+  PL_ASSERT(tile_group_header->GetBeginCommitId(tuple_id) == MAX_CID);
+  PL_ASSERT(tile_group_header->GetEndCommitId(tuple_id) == MAX_CID);
 
   // Add the old tuple into the update set
-  auto old_location = tile_group_header->GetPrevItemPointer(tuple_id);
+  auto old_location = tile_group_header->GetNextItemPointer(tuple_id);
   if (old_location.IsNull() == false) {
-    // if this version is not newly inserted.
+    // update an inserted version
     current_txn->RecordUpdate(old_location);
   }
 }
 
 void TsOrderTxnManager::PerformDelete(const ItemPointer &old_location,
-                                         const ItemPointer &new_location) {
-  auto transaction_id = current_txn->GetTransactionId();
+                                      const ItemPointer &new_location) {
+  LOG_TRACE("Performing Delete");
 
   auto tile_group_header = catalog::Manager::GetInstance()
-                               .GetTileGroup(old_location.block)
-                               ->GetHeader();
+      .GetTileGroup(old_location.block)->GetHeader();
   auto new_tile_group_header = catalog::Manager::GetInstance()
-                                   .GetTileGroup(new_location.block)
-                                   ->GetHeader();
+      .GetTileGroup(new_location.block)->GetHeader();
+
+  auto transaction_id = current_txn->GetTransactionId();
+
+  PL_ASSERT(GetLastReaderCid(tile_group_header, old_location.offset) <= current_txn->GetBeginCommitId());
 
-  // if we can perform update, then we must have already locked the older
-  // version.
   PL_ASSERT(tile_group_header->GetTransactionId(old_location.offset) ==
-            transaction_id);
+         transaction_id);
   PL_ASSERT(new_tile_group_header->GetTransactionId(new_location.offset) ==
-            INVALID_TXN_ID);
+         INVALID_TXN_ID);
   PL_ASSERT(new_tile_group_header->GetBeginCommitId(new_location.offset) ==
-            MAX_CID);
-  PL_ASSERT(new_tile_group_header->GetEndCommitId(new_location.offset) ==
-            MAX_CID);
+         MAX_CID);
+  PL_ASSERT(new_tile_group_header->GetEndCommitId(new_location.offset) == MAX_CID);
 
   // Set up double linked list
-  tile_group_header->SetNextItemPointer(old_location.offset, new_location);
-  new_tile_group_header->SetPrevItemPointer(new_location.offset, old_location);
+  
+  auto old_prev = tile_group_header->GetPrevItemPointer(old_location.offset);
+
```

Comment:
Can you make this logic a separate function ? It looks like this is repeated in different functions.